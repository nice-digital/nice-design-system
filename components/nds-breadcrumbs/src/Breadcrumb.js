import React from "react";
import PropTypes from "prop-types";

export const Breadcrumb = props => {
	const { elementType, to, children, ...attributes } = props;

	const innerTagProps = {
		...attributes
	};

	let InnerHtmlTag = "span";

	if (to) {
		InnerHtmlTag = elementType || "a";
		innerTagProps[InnerHtmlTag === "a" ? "href" : "to"] = to;
	}

	return (
		<li className="breadcrumbs__crumb">
			<InnerHtmlTag {...innerTagProps}>{children}</InnerHtmlTag>
		</li>
	);
};

Breadcrumb.propTypes = {
	elementType: PropTypes.elementType, // Allow tag to be customised for custom routing integration
	children: PropTypes.string.isRequired,
	to: PropTypes.string
};

Breadcrumb.defaultProps = {
	elementType: "a"
};
