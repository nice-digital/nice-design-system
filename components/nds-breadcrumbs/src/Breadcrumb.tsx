import { ReactNode, ElementType } from "react";
import classnames from "classnames";

export interface BreadcrumbProps {
	[prop: string]: unknown;
	children: ReactNode;
	to?: string;
	elementType?: ElementType;
	method?: string;
	className?: string;
}

export const Breadcrumb = (props: BreadcrumbProps) => {
	const { elementType, method, to, children, className, ...attributes } = props;

	let ElementType = elementType || "span";

	const innerTagProps = {
		...attributes
	};

	if (to) {
		ElementType = elementType || "a";
		innerTagProps[method || (ElementType === "a" && "href") || "to"] = to;
	}

	return (
		<li className={classnames(["breadcrumbs__crumb", className])}>
			<ElementType {...innerTagProps}>{children}</ElementType>
		</li>
	);
};
