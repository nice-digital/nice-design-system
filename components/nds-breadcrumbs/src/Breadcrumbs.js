import React from "react";
import PropTypes from "prop-types";

import { Breadcrumb } from "./Breadcrumb";
import "../scss/breadcrumbs.scss";

export { Breadcrumb };

const stringify = (data: any) => {
	if (process.env.NODE_ENV === "production") return JSON.stringify(data);
	return JSON.stringify(data, null, 2);
};

const getBreadcrumbJsonLdItem = (breadcrumb, index = 0) => ({
	"@type": "ListItem",
	position: index + 1,
	item: {
		"@id": breadcrumb.props.to,
		name: breadcrumb.props.children
	}
});

export const Breadcrumbs = ({ children }) => {
	const jsonLdData = {
		"@context": "http://schema.org",
		"@type": "BreadcrumbList",
		itemListElement: children.length
			? children.map(getBreadcrumbJsonLdItem)
			: getBreadcrumbJsonLdItem(children)
	};

	return (
		<>
			<nav aria-label="Breadcrumbs" role="navigation">
				<ol className="breadcrumbs">{children}</ol>
			</nav>
			<script
				type="application/ld+json"
				dangerouslySetInnerHTML={{ __html: stringify(jsonLdData) }}
			/>
		</>
	);
};

Breadcrumbs.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(Breadcrumb),
		PropTypes.objectOf(Breadcrumb)
	]).isRequired
};
