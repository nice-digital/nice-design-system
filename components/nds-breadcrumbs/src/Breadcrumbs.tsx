import { ReactElement, ReactNode } from "react";
import { type BreadcrumbProps } from "./Breadcrumb";

import { Breadcrumb } from "./Breadcrumb";
import "../scss/breadcrumbs.scss";

export { Breadcrumb };

export interface BreadcrumbsProps {
	[prop: string]: unknown;
	className?: string;
	children: ValidBreadcrumbProp[] | ValidBreadcrumbProp;
}

type ValidBreadcrumbProp = ReactElement<BreadcrumbProps> | null | undefined;

const stringify = (data: Object) => {
	if (process.env.NODE_ENV === "production") return JSON.stringify(data);
	return JSON.stringify(data, null, 2);
};

const getBreadcrumbJsonLdItem = (
	breadcrumb: ValidBreadcrumbProp,
	index = 0
) => ({
	"@type": "ListItem",
	position: index + 1,
	item: {
		"@id": breadcrumb?.props.to,
		name: breadcrumb?.props.children
	}
});

export const Breadcrumbs = ({ children, ...rest }: BreadcrumbsProps) => {
	const jsonLdData = {
		"@context": "http://schema.org",
		"@type": "BreadcrumbList",
		itemListElement: Array.isArray(children)
			? children.filter(Boolean).map(getBreadcrumbJsonLdItem)
			: getBreadcrumbJsonLdItem(children)
	};

	return (
		<>
			<nav
				aria-label="Breadcrumbs"
				role="navigation"
				data-component="breadcrumbs"
				{...rest}
			>
				<ol className="breadcrumbs">{children}</ol>
			</nav>
			<script
				type="application/ld+json"
				dangerouslySetInnerHTML={{ __html: stringify(jsonLdData) }}
			/>
		</>
	);
};
