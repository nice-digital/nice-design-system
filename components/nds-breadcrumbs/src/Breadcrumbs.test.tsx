import { render } from "@testing-library/react";

import { Breadcrumb } from "./Breadcrumb";
import { Breadcrumbs } from "./Breadcrumbs";

describe("Breadcrumbs", () => {
	it("should match snapshot for single level", () => {
		const wrapper = render(
			<Breadcrumbs>
				<Breadcrumb to="/">Home</Breadcrumb>
				<Breadcrumb>Page</Breadcrumb>
			</Breadcrumbs>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match snapshot for multiple levels", () => {
		const wrapper = render(
			<Breadcrumbs>
				<Breadcrumb to="/">Home</Breadcrumb>
				<Breadcrumb to="/topics/">Topics</Breadcrumb>
				<Breadcrumb>Page</Breadcrumb>
			</Breadcrumbs>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should spread additional props onto the container", () => {
		const wrapper = render(
			<Breadcrumbs data-track={false}>
				<Breadcrumb to="/">Home</Breadcrumb>
				<Breadcrumb>Page</Breadcrumb>
			</Breadcrumbs>
		);
		expect(wrapper.getByRole("navigation").getAttribute("data-track")).toBe(
			"false"
		);
	});
});
