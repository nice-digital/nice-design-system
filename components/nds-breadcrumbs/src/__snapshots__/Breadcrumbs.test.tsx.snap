// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should match snapshot for multiple levels 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumbs"
        data-component="breadcrumbs"
        role="navigation"
      >
        <ol
          class="breadcrumbs"
        >
          <li
            class="breadcrumbs__crumb"
          >
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="breadcrumbs__crumb"
          >
            <a
              href="/topics/"
            >
              Topics
            </a>
          </li>
          <li
            class="breadcrumbs__crumb"
          >
            <span>
              Page
            </span>
          </li>
        </ol>
      </nav>
      <script
        type="application/ld+json"
      >
        {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/topics/",
        "name": "Topics"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Page"
      }
    }
  ]
}
      </script>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumbs"
      data-component="breadcrumbs"
      role="navigation"
    >
      <ol
        class="breadcrumbs"
      >
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/topics/"
          >
            Topics
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <span>
            Page
          </span>
        </li>
      </ol>
    </nav>
    <script
      type="application/ld+json"
    >
      {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/topics/",
        "name": "Topics"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Page"
      }
    }
  ]
}
    </script>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Breadcrumbs should match snapshot for single level 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumbs"
        data-component="breadcrumbs"
        role="navigation"
      >
        <ol
          class="breadcrumbs"
        >
          <li
            class="breadcrumbs__crumb"
          >
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="breadcrumbs__crumb"
          >
            <span>
              Page
            </span>
          </li>
        </ol>
      </nav>
      <script
        type="application/ld+json"
      >
        {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "name": "Page"
      }
    }
  ]
}
      </script>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumbs"
      data-component="breadcrumbs"
      role="navigation"
    >
      <ol
        class="breadcrumbs"
      >
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <span>
            Page
          </span>
        </li>
      </ol>
    </nav>
    <script
      type="application/ld+json"
    >
      {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "name": "Page"
      }
    }
  ]
}
    </script>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
